Initial support for collision handling

andresilva opened the pull request:
This PR adds initial support for collision handling. A `compact()` method was added that scans through the data file checking for prefixes that have more than `N` collisions. Key-value pairs with collided prefixes are moved to their own file. Information about collided prefixes is stored in the metadata in its own `PrefixTree`, this way when performing a `get` we can jump to the appropriate collision file directly.

The current implementation of the collision file is an append-only log extended with an in-memory index (this means that all keys for a given collision file are currently stored in-memory). This can be improved afterwards (e.g. mmap log, replace with B+Tree, or use the same strategy we use for the existing data file but using other bits for the prefix calculation).

Currently, flushing collided prefixes is done right away, no virtual commit is made beforehand.

**TODO:**
- [x] Extend collision file with in-memory index
- [x] Iterate through collision files
- [x] Make the maximum number of collisions configurable
- [x] Tests
- [x] Docs
- [x] Clean up

andresilva commented:
I'm not sure whether we should be controlling for prefix collisions or actual key displacement from their expected position. Either way this should be easy to update afterwards.

Please don't waste your time reviewing this yet as there's still a lot of stuff that needs cleaning.

andresilva commented:
I think this is ready for a review. Some general thoughts:
* Data returned from the collision file is currently owned (I have some ideas of how we can improve this)
https://github.com/paritytech/paritydb/blob/e59efc067aa50817e1100bcfa6d1a1f3f19ce93a/paritydb/src/collision.rs#L20-L25
* There is a failing test (that I added) since it depends on the fix from #63.
* The compaction process is currently copying all keys to memory. This may be fine for now but should be improved in the future (I couldn't find an easy way to make the borrow checker happy)
* We should probably refactor our `Value` type (#39)
* I would prefer it if there was more encapsulation/isolation of the three different data structures that make up the database: journal, data file and collision files
  * For example, right now there is no first-class data structure to represent the data file (it exists as a `Mmap` in `Database` and is passed around as a `&[u8]` when used. 
  * There are already some issues somewhat related to this (#16, #23)
* I'm not particularly fond of the naming I came up with (`struct Collision`), I'm open to suggestions
* I didn't add any tests for `Metadata` since we currently have a PR open for that (#53) which we can extend afterwards

andresilva commented:
The test failed but the build passed, should be fixed with #64.

tomusdrw commented on paritydb/src/database.rs:
```diff
@@ -231,19 +285,150 @@ impl Database {
+		let mut collided_prefixes = Vec::new();
+		let mut collision_files = Vec::new();
+
+		for (prefix, keys) in collisions.iter() {
```
To avoid over-allocation and traversing entire database we can do it dynamically:
1. Read all keys for a prefix
2. Perform compaction if the value is greater.

tomusdrw commented on paritydb/src/database.rs:
```diff
@@ -231,19 +285,150 @@ impl Database {
+			flush.delete()?;
+		}
+
+		// update metadata
```
This is not atomic, might be better to do it other way around:
1. Update collisions and update+flush metadata mmap
2. Create the flush
3. Perform the flush

This way if it fails in the middle the flush will be applied anyway and metadata will be propelry updated if I'm not mistaken.

andresilva replied:
I think it should be instead:

1. Update collisions
2. Create the flush
3. Update and flush metadata mmap
4. Perform the flush

If we fail during the first step, the collision file is ignored. After the flush is properly created we can safely update metadata, since we now that the flush will be reapplied on restart.

andresilva commented on paritydb/src/collision.rs:
```diff
@@ -0,0 +1,323 @@
+/// bits of the prefix and adding extra bits as needed
+///
+#[derive(Debug)]
+pub struct Collision {
```
Yes, I will implement this in this PR.

tomusdrw replied:
IMHO better to leave it for a separate PR, this one is getting big already. Just be sure to create a separate issue for this.

andresilva commented on paritydb/src/collision.rs:
```diff
@@ -0,0 +1,323 @@
+				.open(&self.path)?;
+
+			let mut file = BufReader::new(file);
+			file.seek(SeekFrom::Start(entry.position))?;
```
The collision files are cached in `Database`, so step 1 and 2 should only happen once. But I agree, I will refactor this to use `mmap`. In order to avoid reopening the `mmap` on every write/delete we could grow the file in chunks, but I will implement the simpler version for now.

andresilva commented on paritydb/src/collision.rs:
```diff
@@ -0,0 +1,323 @@
+
+impl<'a> CollisionLogIterator<'a> {
+	fn new<P: AsRef<Path>>(path: P, index_iter: btree_map::Values<'a, Vec<u8>, IndexEntry>)
+						   -> Result<CollisionLogIterator<'a>> {
```
I agree it's ugly. We currently don't have a `rustfmt.toml`, so it's on me.

```
	fn new<P: AsRef<Path>>(
		path: P,
		index_iter: btree_map::Values<'a, Vec<u8>, IndexEntry>
	) -> Result<CollisionLogIterator<'a>> {
```

Should I format it like this instead?

tomusdrw replied:
I prefer the latter tbh (but with a comma after second argument :))

andresilva commented on paritydb/src/collision.rs:
```diff
@@ -0,0 +1,323 @@
+	fn read<R: Read + Seek>(reader: &mut R) -> io::Result<LogEntry> {
+		let position = reader.seek(SeekFrom::Current(0))?;
+		let key_size = reader.read_u32::<LittleEndian>()?;
+		let mut key = vec![0u8; key_size as usize];
```
paritydb has fixed key sizes and (possibly) value sizes but the collision file is currently not taking advantage of that. Can we leave this as it is for now, and in another PR I will make the collision file enforce key size (so we won't even need to store it on the disk)?

andresilva commented on paritydb/src/database.rs:
```diff
@@ -231,19 +285,150 @@ impl Database {
+
+		let records = self.record_iter()?;
+
+		Ok(Box::new(records.merge_join_by(collided_records, |r, c| {
```
`interleaved_ordered` requires both iterators to implement `Ord`. This would require implementing `Ord` for both raw values and record values (which might not "fit" in a single slice). We have an issue open for refactoring the `Value` type (#39), maybe we can replace this with `interleaved_ordered` in another PR when we close that issue?

tomusdrw replied:
Fine with me, although the `Ord` implementation should basically just delegate to `key` (we don't really want to compare values)

andresilva commented on paritydb/src/database.rs:
```diff
@@ -181,22 +203,46 @@ impl Database {
+						"prefix is declared as collided; \
+						 collision file should exist in collisions index; qed");
+
+					collision.apply(op)?;
```
Yes, this can be cleaned up afterwards with log compaction. Also, the collision log now avoids inserts for same values.

andresilva commented on paritydb/src/options.rs:
```diff
@@ -64,6 +66,7 @@ impl Default for Options {
 			key_index_bits: 8,
 			key_len: 32,
 			value_len: ValuesLen::Constant(64),
+			max_prefix_collisions: 6,
```
This value is arbitrary, it should probably be related to `key_index_bits`.

tomusdrw replied:
I think it's fine if it's constant (but configurable), how do you think it should be related? Bigger database does not necessarily mean that we can handle more collisions.

andresilva commented:
@tomusdrw I think I have addressed all your comments. I created/updated issues for things that I think could be done on other PRs (#67, #66, #65, #39). 

I have rewritten the compaction algorithm to perform compactions incrementally by prefix, it no longer copies keys to memory, and I think it should be crash-safe now. Unfortunately it requires some unsafety (aliasing raw pointers) since we need to mutate the database while iterating over its records. Let me know if you have any ideas to avoid this.

tomusdrw commented on paritydb/src/collision.rs:
```diff
@@ -228,52 +297,61 @@ impl LogEntry {
-		let value_size = reader.read_u32::<LittleEndian>()?;
+	// FIXME: should return Result
+	fn read(data: &[u8]) -> (usize, LogEntry) {
+		// FIXME: add sanity limits for key_size and value_size
```
Alternatively add checksum at the beginning of the file (imho preferrable)

tomusdrw commented on paritydb/src/database.rs:
```diff
@@ -345,90 +344,115 @@ impl Database {
+						self.metadata_mmap.flush()?;
+					}
+
+					commits.push((collision, flush));
```
I don't think we support many flush files. Why not create the collision file and flush the database after each key?

I know it will probably be more expensive, but I don't expect it to be frequent operation. Perhaps `compact` could get a `Range` parameter to perform compaction only on given part of database if necessary? Or a `max_compactions` parameter to limit the time required for compaction?

tomusdrw replied:
Perhaps revert that commit, and prepare a better compaction in a separate PR? I'm ok with previous implementation (if it works, might be good to have a test for that), while I don't think this one would work.

tomusdrw commented on paritydb/src/database.rs:
```diff
@@ -345,90 +344,115 @@ impl Database {
+			// create a new collision file and create a flush to delete the moved data (without
+			// applying it).
+			let mut handle_prefix_compaction = |prefix: u32, keys: &[&[u8]]| -> Result<()> {
+				if keys.len() >= self.options.external.max_prefix_collisions {
```
Might be better to exit early instead indenting the whole code.

tomusdrw commented on paritydb/src/database.rs:
```diff
@@ -345,90 +344,115 @@ impl Database {
 	pub fn compact(&mut self) -> Result<Vec<u32>> {
-		let collisions = self.collisions()?;
+		let mut current_prefix = None;
+		let mut prefix_keys: Vec<&[u8]> = Vec::new();
```
might be good to allocate `with_capacity(max_collisions)` I would also move the declaration down where it's needed (close to the for loop)

tomusdrw commented on paritydb/src/collision.rs:
```diff
@@ -0,0 +1,409 @@
+		let mut index = BTreeMap::new();
+
+		for (position, entry) in log {
+			match entry.value {
```
Might look a bit better (less indentation) with:
```
if let Some(value) = .. {

} else {

}
```

tomusdrw commented on paritydb/src/options.rs:
```diff
@@ -99,6 +102,13 @@ impl InternalOptions {
 			));
 		}
 
+		if external.max_prefix_collisions < 1 {
```
Actually it needs to be greater than 1, otherwise we will move single values to collision files.

andresilva replied:
It's a good way to benchmark the performance of the collision files 😛

tomusdrw commented on paritydb/src/database.rs:
```diff
@@ -231,19 +286,174 @@ impl Database {
+						// FIXME: store a reference to the value in the `keys` vec
+						let value = self.get(key)?.expect("The key has been returned by the iterator; qed");
+						// FIXME: only need to copy the value if it isn't stored in a single field
+						collision.insert(key, &value.to_vec())?;
```
Would be best to never copy the value (we don't really want to or need to load it into memory), why not accept `Value` (or a refactored version of it) inside `Collision`?

andresilva commented:
I've reverted the latest changes to the compaction algorithm (created #69), and I've updated the compaction algorithm so that it doesn't copy data and I think it should be crash-safe now.

tomusdrw commented on paritydb/src/database.rs:
```diff
@@ -231,19 +295,152 @@ impl Database {
+				for key in keys {
+					// FIXME: store a reference to the value in the return Map from collisions
+					let value = self.get(&key)?.expect("The key has been returned by the iterator; qed");
+					collision_file.insert(&key, value.as_slice().unwrap_or(&value.to_vec()))?;
```
Still may allocate, please refactor when #39 is being done.

tomusdrw commented:
Looks good! Ready to merge imho. @debris please have a look (note that some grumbles will be addressed in separate PRs)

debris commented:
nice!
